problem_builder/mixins.py

```diff
@@ -221,3 +238,57 @@ def student_view_user_state(self, context=None, suffix=''):
         json_result = json.dumps(result, cls=DateTimeEncoder)
 
         return webob.response.Response(body=json_result, content_type='application/json')
+
+
+class StudentViewUserStateResultsTransformerMixin(object):
+    """
+    A convenient way for MentoringBlock and MentoringStepBlock to share
+    student_results transform code.
+    Needs to be used alongside StudentViewUserStateMixin.
+    """
+    def transforms(self):
+        return {
+            'student_results': self.transform_student_results
+        }
+
+    def student_results_transforms(self):
+        """
+        Key (inside `student_results`) to transform function map.
+        Should take
+        """
+        return [
+            ('choices', self.transform_choices),
+            ('tips', self.transform_tips),
+        ]
+
+    def transform_student_results(self, student_results):
+        """
+        Applies `student_results_transforms()` output to current `student_results`
+        """
+        current_student_results = dict(student_results)
+        transformed_student_results = []
+        for question in self.steps:
+            if question.name in current_student_results:
+                question_results = question.get_results(current_student_results[question.name])
+                for key, f in self.student_results_transforms():
+                    if key in question_results:
+                        question_results[key] = f(question_results, question)
+                transformed_student_results.append([question.name, question_results])
+        return transformed_student_results
+
+    def transform_tips(self, result, question):
+        """
+        For MCQ, tips are at `student_results` root level.
+        """
+        return [tip.student_view_data() for tip in question.get_tips() if question.student_choice in tip.values]
```

Comment:
Actually, as mentioned elsewhere, these tips are not user-specific so I think we should only include them in `student_view_data`, not also here in the `user_state`.

Reply:
@bradenmacdonald  Is it not enough that we only output tips that are linked to the user's choice here?

Reply:
@bradmerlin  Yeah, it's not really a problem, but it doesn't seem useful to me, since the API consumer will already know what the relevant tips are, since it knows all the tips and it knows the learner's selected answer. So I think this is redundant.